src/regexp.js

```diff
@@ -0,0 +1,1124 @@
+const LEFT_PARENTHESIS = 0x28 // (
+const RIGHT_PARENTHESIS = 0x29 // )
+const ASTERISK = 0x2A // *
+const PLUS_SIGN = 0x2B // +
+const COMMA = 0x2C // ,
+const HYPHEN_MINUS = 0x2D // -
+const FULL_STOP = 0x2E // .
+const SOLIDUS = 0x2F // /
+const DIGIT_ZERO = 0x30 // 0
+const DIGIT_ONE = 0x31 // 1
+const DIGIT_SEVEN = 0x37 // 7
+const DIGIT_NINE = 0x39 // 9
+const COLON = 0x3A // :
+const LESS_THAN_SIGN = 0x3C // <
+const EQUALS_SIGN = 0x3D // =
+const GREATER_THAN_SIGN = 0x3E // >
+const QUESTION_MARK = 0x3F // ?
+const LATIN_CAPITAL_LETTER_A = 0x41 // A
+const LATIN_CAPITAL_LETTER_B = 0x42 // B
+const LATIN_CAPITAL_LETTER_D = 0x44 // D
+const LATIN_CAPITAL_LETTER_F = 0x46 // F
+const LATIN_CAPITAL_LETTER_P = 0x50 // P
+const LATIN_CAPITAL_LETTER_S = 0x53 // S
+const LATIN_CAPITAL_LETTER_W = 0x57 // W
+const LATIN_CAPITAL_LETTER_Z = 0x5A // Z
+const LOW_LINE = 0x5F // _
+const LATIN_SMALL_LETTER_A = 0x61 // a
+const LATIN_SMALL_LETTER_B = 0x62 // b
+const LATIN_SMALL_LETTER_C = 0x63 // c
+const LATIN_SMALL_LETTER_D = 0x64 // d
+const LATIN_SMALL_LETTER_F = 0x66 // f
+const LATIN_SMALL_LETTER_K = 0x6B // k
+const LATIN_SMALL_LETTER_N = 0x6E // n
+const LATIN_SMALL_LETTER_P = 0x70 // p
+const LATIN_SMALL_LETTER_R = 0x72 // r
+const LATIN_SMALL_LETTER_S = 0x73 // s
+const LATIN_SMALL_LETTER_T = 0x74 // t
+const LATIN_SMALL_LETTER_U = 0x75 // u
+const LATIN_SMALL_LETTER_V = 0x76 // v
+const LATIN_SMALL_LETTER_W = 0x77 // w
+const LATIN_SMALL_LETTER_X = 0x78 // x
+const LATIN_SMALL_LETTER_Z = 0x7A // z
+const LEFT_SQUARE_BRACKET = 0x5B // [
+const REVERSE_SOLIDUS = 0x5C // \
+const RIGHT_SQUARE_BRACKET = 0x5D // [
+const CIRCUMFLEX_ACCENT = 0x5E // ^
+const LEFT_CURLY_BRACKET = 0x7B // {
+const VERTICAL_LINE = 0x7C // |
+const RIGHT_CURLY_BRACKET = 0x7D // }
+const ZERO_WIDTH_NON_JOINER = 0x200C
+const ZERO_WIDTH_JOINER = 0x200D
+
+const pp = Parser.prototype
+
+export class RegExpValidationState {
+  constructor(parser) {
+    this.parser = parser
+    this.validFlags = `gim${parser.options.ecmaVersion >= 6 ? "uy" : ""}${parser.options.ecmaVersion >= 9 ? "s" : ""}`
+    this.source = ""
+    this.flags = ""
+    this.start = 0
+    this.switchU = false
+    this.switchN = false
+    this.pos = 0
+    this.lastIntValue = 0
+    this.lastStringValue = ""
+    this.lastAssertionIsQuantifiable = false
+    this.numCapturingParens = 0
+    this.maxBackReference = 0
+    this.groupNames = []
+    this.backReferenceNames = []
+  }
+
+  reset(start, pattern, flags) {
+    const unicode = flags.indexOf("u") !== -1
+    this.start = start | 0
+    this.source = pattern + ""
+    this.flags = flags
+    this.switchU = unicode && this.parser.options.ecmaVersion >= 6
+    this.switchN = unicode && this.parser.options.ecmaVersion >= 9
+  }
+
+  raise(message) {
+    this.parser.raise(this.start, `Invalid regular expression: /${this.source}/: ${message}`)
+  }
+
+  // Node.js 0.12/0.10 don't support String.prototype.codePointAt().
+  codePointAt(i) {
```

Comment:
Is there any case where it actually matters whether we have a single or two (non-syntax) characters? Could we make do by just looking at code units not code points?

(If not, there may be a problem where this gets regexps without `u` flag wrong, since those _should_ treat surrogate pairs as two characters. If yes, we could maybe drop the extra complexity around code points?)

Reply:
Thank you for good catch. It was a problem. I fixed it and added tests.